Pass source ranges parameters by move inside range constructors

nordlow opened the pull request:
This enables non-copyable (r-value) ranges to be passed to a couple of range algorithms. My motivation for this is to enable non-copyable values, typically containers, to be passed by move into a specific range I currently call `UniqueRange`. `UniqueRange` being, in turn non-copyable, should then be passed to Phobos' higher-order algorithms, typically map, filter, reduce etc. This is very useful when returning, for instance, containers combined with map, filter and other higher-order functions. This has the same motivation as Rusts' `[into_iter()](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html)`.

Here's a sample implementation of this idea that compiles and tests out correctly with this branch of Phobos.

https://github.com/nordlow/phobos-next/blob/master/src/unique_range.d

being a wrapper for

https://github.com/nordlow/phobos-next/blob/master/src/array_ex.d

Note, however, that this fix fails a unittest. 

IMO the best way to solve this is to fix the compiler so that it automatically converts the last reference of a value-passes parameter to an r-value (probably involving calling `doCopyOrMove()` to the correct DMD's ast node somewhere `src/expression.d`). Typically the assignment in the range constructor. But I haven't figure out how to do this. I have and idea outlined here: https://github.com/nordlow/dmd/commit/d3053e7801e340541bba32172718c2bb3de966c0 but I need feedback from the kernel developers on how go forward with this if you think that is a better way of solving this. Specifically the visitor that finds the last reference of a symbol that can be passed by move (non-reference) parameter in current function (scope).

All this enables D to compete with Rust by enabling range-algorithms to be safely combined with non-copyable containers. This is especially useful when returning lazy-computation in function return values.

andralex commented:
This should be automatically done by the compiler. If a by-value argument is only used once, it should be automatically moved. cc @WalterBright

nordlow commented:
Except for

```D
static if (isCopyable!Range)
{
    auto opSlice() { return this; }
}
```

AFAICT.

andralex commented:
@nordlow yes, and probably with `.save`, too

nordlow commented:
Should I added these now?

andralex commented:
@nordlow that would be great. Perhaps in a different PR to keep this as evidence for the language issue.

nordlow commented:
So you like the general idea here, @andralex ?

nordlow commented:
Ok, I'll add these checks now.

andralex commented:
@nordlow yes, we should accept more uses of uncopyable types in the core language and stdlib. thx!

MetaLang commented:
@nordlow can you either rebase or close this PR?

nordlow commented:
As mentioned above, @andralex thinks we should fix the compiler first before pulling this. I haven't gotten any assistance on fixing the compiler, though.

Could somebody please help me move forward with my very very unfinished idea at https://github.com/nordlow/dmd/commit/d3053e7801e340541bba32172718c2bb3de966c0 that makes D's move semantics kick in in more places?

RazvanN7 commented:
@nordlow Could you maybe report the compiler issue on bugzilla? I can look into it if a have a clear reproducible test case.

nordlow commented:
> @nordlow Could you maybe report the compiler issue on bugzilla? I can look into it if a have a clear reproducible test case.

Very happy to see interest in this being resurrected! I'll make an issue. Btw, we should have a talk with @WalterBright about this because he has mentioned interest in relaxing the rules for move aswell. Ping, @andralex.

nordlow commented:
Issue at https://issues.dlang.org/show_bug.cgi?id=22396, @RazvanN7.

nordlow commented:
> @nordlow Now I understand what the problem was. I have commented on the issue also: this will probably be fixed once DIP1040 is approved and implemented

Ok, thanks. I guess we should tag this as blocked until the DIP has been approved and implemented then.

nordlow commented:
In the mean-time we could add a single call to move for theLet's forget about this for now. Turns out we cannot currently implement this a generic way thanks to dmd's constructor data-flow analysis. For instance,

```d
    this(R input)
    {
        import std.algorithm.mutation : moveEmplace;
        () @trusted { moveEmplace(input, _input); } ();
    }
```

fails to compile as

```
std/algorithm/iteration.d(558): Error: field `_input` must be initialized in constructor, because it is nested struct
```

. We should stop promoting incomplete diagnostics to the compiler that limits expression, @andralex @maxhaton.
